hoomd/md/ManifoldCylinder.h

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2009-2021 The Regents of the University of Michigan
+// This file is part of the HOOMD-blue project, released under the BSD 3-Clause License.
+
+
+// Maintainer: pschoenhoefer
+
+#ifndef __MANIFOLD_CLASS_CYLINDER_H__
+#define __MANIFOLD_CLASS_CYLINDER_H__
+
+#include "hoomd/HOOMDMath.h"
+#include "hoomd/BoxDim.h"
+#include <pybind11/pybind11.h>
+
+/*! \file ManifoldCylinder.h
+    \brief Defines the manifold class for the Cylinder surface
+*/
+
+// need to declare these class methods with __device__ qualifiers when building in nvcc
+// DEVICE is __host__ __device__ when included in nvcc and blank when included into the host compiler
+#ifdef __HIPCC__
+#define DEVICE __device__
+#else
+#define DEVICE
+#endif
+
+//! Class for constructing the Cylinder surface
+/*! <b>General Overview</b>
+
+    ManifoldCylinder is a low level computation class that computes the distance and normal vector to the Cylinder surface.
+
+    <b>Cylinder specifics</b>
+
+    ManifoldCylinder constructs the surface:
+    R^2 = (x-P_x)^2 + (y-P_y)^2
+
+    These are the parameters:
+    - \a P_x = center position of the cylinder in x-direction;
+    - \a P_y = center position of the cylinder in y-direction;
+    - \a P_z = center position of the cylinder in z-direction;
+    - \a R = radius of the cylinder;
+
+*/
+
+class ManifoldCylinder
+    {
+    public:
+
+        //! Constructs the manifold class
+        /*! \param _P center position in x-y-z coordinates
+            \param _R radius
+        */
+        DEVICE ManifoldCylinder(const Scalar _R, const Scalar3 _P)
+            : Px(_P.x), Py(_P.y), Pz(_P.z), R_sq(_R*_R)
+            {
+            }
+
+        //! Evaluate implicit function
+        /*! \param point Point at which surface is calculated
+
+            \return result of the nodal function at input point
+        */
+
+        DEVICE Scalar implicitFunction(const Scalar3& point)
+        {
+            return  (point.x - Px)*(point.x - Px) + (point.y - Py)*(point.y - Py) - R_sq;
+        }
+
+        //! Evaluate deriviative of implicit function
+        /*! \param point Point at surface is calculated
+
+            \return normal of the Cylinder surface at input point
+        */
+
+        DEVICE Scalar3 derivative(const Scalar3& point)
+        {
+            return make_scalar3(2*(point.x - Px), 2*(point.y - Py), 0);
+        }
+
+        DEVICE bool fitToBox(const BoxDim& box)
```

Comment:
I still find it very strange that a method named with an action verb (`fit`) is defined such that sometimes it modifies state and sometimes it doesn't, but it always also returns a boolean condition. I think we've changed the name multiple times over the course of this PR and don't want to belabor the review process any further, but could you add docstrings for the methods so that it's obvious what it's doing? In the future we'll want to document this in ARCHITECTURE.md (see #979), but for now a docstring here would be sufficient.

Reply:
I made a last attempt and now called it `fitsInsideBox` to make clear that this checks if the manifold can fit inside the specified box.

Reply:
That's fine. Good enough for me, naming isn't _that_ important to hold this up any further 😂